### ISSUE TITLE
PyPI release binary wheel unsupported

### ISSUE THREAD

**Issue by cspencerjones** [cspencerjones/xlayers#5]
Perhaps @andersy005 can help me with this? I am trying to upload the xlayers package to PyPI using the instructions here: https://github.com/xgcm/xgcm/issues/168#issuecomment-564645248. But when I run 
`twine upload dist/*`, 

I get the error
`HTTPError: 400 Client Error: Binary wheel 'xlayers-0.1-cp27-cp27mu-linux_x86_64.whl' has an unsupported platform tag 'linux_x86_64'`

I googled, but I'm having trouble decoding the results - is there a way around this problem? Would it help to build the package on a different machine, or is this problem caused by the xlayers code?

**Comment by andersy005** [comment]
@cspencerjones, my speculation here is that this error happens because PyPi requires [manylinux wheels](https://github.com/pypa/manylinux) wheels. I've never built a package that required `manylinux wheels` before :(. 

If I were in your shoes, I would consider building the package without wheels (I am being lazy here 😀)

```bash
# build package without wheels. Make sure to delete the old sdist directory
python setup.py sdist  
twine upload dist/*  
```
As a result, the user would compile the source code on their machine. They would need the appropriate compilers and libraries for your Python package though. In other words, `pip install package_name` would involve a compilation step. This can be non-trivial if your package is complex or the compilation step is non-trivial (which is why wheels are awesome since the user doesn't need to compile the package themselves). 

Here's a great demo: https://github.com/pypa/python-manylinux-demo that shows how to build *manylinux* wheels. 

Let me know if the demo is useful. I am happy to help out (it might be a good opportunity for me to learn how to build *manylinux* wheels as well)

**Comment by andersy005** [comment]
@cspencerjones, upon further investigations, I found out that there is an **easy** way to build *manylinux* wheels. I will post a detailed update tomorrow.

**Comment by andersy005** [comment]
I was able to get [cibuildwheel package](https://cibuildwheel.readthedocs.io/en/stable/) to build wheels for xlayers:

- Linux:

```bash
$ ls -ltrh wheelhouse
total 5.5M
-rw-r--r-- 1 runner docker 819K Dec 17 22:09 xlayers-0.1+16.gf74f8ef-cp38-cp38-manylinux2010_i686.whl
-rw-r--r-- 1 runner docker 175K Dec 17 22:07 xlayers-0.1+16.gf74f8ef.tar.gz
-rw-r--r-- 1 runner docker 992K Dec 17 22:08 xlayers-0.1+16.gf74f8ef-cp36-cp36m-manylinux2010_x86_64.whl
-rw-r--r-- 1 runner docker 995K Dec 17 22:08 xlayers-0.1+16.gf74f8ef-cp37-cp37m-manylinux2010_x86_64.whl
-rw-r--r-- 1 runner docker 995K Dec 17 22:08 xlayers-0.1+16.gf74f8ef-cp38-cp38-manylinux2010_x86_64.whl
-rw-r--r-- 1 runner docker 816K Dec 17 22:09 xlayers-0.1+16.gf74f8ef-cp36-cp36m-manylinux2010_i686.whl
-rw-r--r-- 1 runner docker 819K Dec 17 22:09 xlayers-0.1+16.gf74f8ef-cp37-cp37m-manylinux2010_i686.whl
-rw-r--r-- 1 runner docker 819K Dec 17 22:09 xlayers-0.1+16.gf74f8ef-cp38-cp38-manylinux2010_i686.whl
```

- MacOS

```bash
$ ls -ltrh wheelhouse
total 7120
-rw-r--r--  1 runner  staff   174K Dec 17 22:07 xlayers-0.1+16.gf74f8ef.tar.gz
-rw-r--r--  1 runner  staff   1.1M Dec 17 22:09 xlayers-0.1+16.gf74f8ef-cp36-cp36m-macosx_10_6_intel.whl
-rw-r--r--  1 runner  staff   1.1M Dec 17 22:10 xlayers-0.1+16.gf74f8ef-cp37-cp37m-macosx_10_6_intel.whl
-rw-r--r--  1 runner  staff   1.1M Dec 17 22:11 xlayers-0.1+16.gf74f8ef-cp38-cp38-macosx_10_9_x86_64.whl
```

As stated in the cibuildwheel documentation:
> cibuildwheel is not intended to run on your development machine. Because it uses system Python from Python.org it will try to install packages globally - not what you expect from a build tool! Instead, isolated CI services like Travis CI, CircleCI, Azure Pipelines and AppVeyor are ideal.

I decided to try out cibuildwheel via the new Github Workflow (CI) feature. The logs are available [here](https://github.com/andersy005/xlayers/commit/f74f8ef1e0677c0549c8cef677af9998e65acff7/checks?check_suite_id=363533280). The configuration for this workflow resides in https://github.com/andersy005/xlayers/blob/master/.github/workflows/pythonpublish.yaml. 


Note:

To upload (automatically) these wheels to PyPI, the repo admin will need to 
add `PYPI_USERNAME` and `PYPI_PASSWORD` secrets in https://github.com/cspencerjones/xlayers/settings/secrets 


@cspencerjones, I am happy to submit a PR if this new feature is of interest.

**Comment by andersy005** [comment]
P.S.:

I am building wheels for Python 3.6 and above. In case support for other versions is needed, `CIBW_SKIP: "cp27-* cp34-* cp35-*"` needs to be updated. 

``` yaml
 - name: Build and publish
        env:
          CIBW_BEFORE_BUILD: "pip install -U numpy"
          CIBW_SKIP: "cp27-* cp34-* cp35-*"
         
        #   TWINE_USERNAME: ${{ secrets.PYPI_USERNAME }}
        #   TWINE_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
        run: |
          python setup.py sdist -d wheelhouse
          python -m cibuildwheel --platform macos --output-dir wheelhouse
```

**Comment by cspencerjones** [comment]
Thanks so much @andersy005 ! If you can submit a PR that would be amazing. I think python 3.6+ is fine. 

I am in the UK for the next couple of weeks, so I may not respond super promptly, but I will get to it!

**Comment by andersy005** [comment]
We can close this issue since it was addressed via #6

**Comment by cspencerjones** [comment]
Thanks @andersy005
